### PR TITLE
Avoid caching textures before they load

### PR DESCRIPTION
**Pull request by zzmp**
- Adds a method to resources denoting whether they should be cached: `Resource::isCacheable`.
- Only sets textures as cacheable once they have both downloaded and processed their image.

Textures had an edge case where they could download binary data, begin processing, go out of scope, and thus never finish processing the data, though it would be cached and so never retried. This fixes that.

Textures now only cache if they have both downloaded *and* processed their data. If they have downloaded but *not processed* their data, they will not be cached. They downloaded binary will be cached according to that file's conventions (asset, file, or HTTP).

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: cb060276d82fdd0651d56503dcff0b01b9f9fa73 available at: https://gist.github.com/37d7995bd4d3be8f501d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7282.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7282.dmg

File: libraries/model-networking/src/model-networking/TextureCache.h [hunk 1]
```diff
@@ -132,6 +132,8 @@ class NetworkTexture : public Resource, public Texture {
 
 protected:
 
+    virtual bool isCacheable() const override { return isCacheable; }
```
Comment (hyperlogic) [inline libraries/model-networking/src/model-networking/TextureCache.h]: whoops, I'm surprised this isn't a warning, the compiler let us down here.  This should be:

    return _isCachable;

isCachable is a pointer to the method, not a bool.

Comment (zzmp) [inline libraries/model-networking/src/model-networking/TextureCache.h]: Weird...it *did* fail on my machine. I must not have pushed up the latest.

**Comment by cozza13** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 883d4ba006faefc08b253ed7a651098d2d0ac548 available at: https://gist.github.com/37d7995bd4d3be8f501d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7282.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7282.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 883d4ba006faefc08b253ed7a651098d2d0ac548 available at: https://gist.github.com/37d7995bd4d3be8f501d<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7282.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7282.dmg

**Comment by bluetoe** [comment]
I do not notice any difference between production build and this one. Not sure how to test to see if it works though.

**Comment by zzmp** [comment]
@bluetoe, use the testing instructions from #7223.

https://app.asana.com/0/74051501495175/95025056969423

Testing

This is a concurrency issue, and is hard to test :smile_cat:. I reproduced it using a debugger to prevent certain lines from executing.

Go to a domain and remove everything in the entity list in edit.js.
Add back a single entity with a texture (https://hifi-public.s3.amazonaws.com/sam/models/mitsuba/mitsuba%20reUVmapped_with%20phong.fbx).
Position yourself to view this entity.
Close interface.
// On a build of master

Set a breakpoint in TextureCache.cpp before calling QThreadPool::globalInstance->start.
Open interface, and skip all occurences of that call.
Remove the breakpoint.
Navigate to a different domain and back. This will move the resource to the _unusedResources cache.
Observe the entity with a still unloaded texture.
// On a build of this PR

Repeat.
Observe the entity with a loaded texture.

**Comment by hyperlogic** [comment]
whoops, I added the wrong tag.

**Comment by OmegaHeron** [comment]
Tested Windows 7 using method from previous PR.  It behaves as expected in that forcing it to miss a texture leads to an eventual correction where it loads texture proper.  It's clear this is improving to me as it's much harder to get it to miss a texture now.  QA Pass

**Pull request merged**